### PR TITLE
Feature/sg 000 introduce class agnoistic nms option

### PR DESCRIPTION
**Pull request by BloodAxe**
- Rename YoloPostPredictionCallback to YoloXPostPrediction callback to stress out that this class is indended for use with YoloX
- Added `class_agnostic_nms` and `multi_label_per_box` parameters (default value corresponds to existing behavior of the callback)

### CODE REVIEW

**Review by shaydeci — CHANGES_REQUESTED** [review]
There is something I might be missing here...

I am a bit confused about how the two arguments multi_label_per_box work together.

If we use multi_label_per_box=True, we pass class indices from the same class that are above the confidence threshold to batched_nms. So

File: src/super_gradients/training/utils/detection_utils.py [hunk 1]
```diff
@@ -242,18 +242,26 @@ def box_area(box):
...
-                                    (instead of the maximum confidence all confidences above threshold
-                                    will be sent to NMS); by default is set to True
-        :param with_confidence: whether to multiply objectness score with class score.
-                                usually valid for Yolo models only.
-        :return: detections with shape nx6 (x1, y1, x2, y2, object_conf, class_conf, class)
+
+    :param prediction: raw model prediction. Should be a list of Tensors of shape (cx, cy, w, h, confidence, cls0, cls1, ...)
+    :param conf_thres: below the confidence threshold - prediction are discarded
+    :param iou_thres: IoU threshold for the nms algorithm
+    :param multi_label_per_box: whether to use re-use each box with all possible labels
```
Comment (shaydeci) [inline src/super_gradients/training/utils/detection_utils.py:254]: The class_agnostic_nms alongside this way of phrasing multi_label_per_box docs might be a bit confusing.
To me this is a bit more clear:
multi_label_per_box: When True, a box with any class confidences > conf_thres will be passed to the NMS algorithm as multiple boxes (with the same coordinates) with different labels. When False, only a single box will be passed to the NMS algorithm with the label of the highest score.

**Review by shaydeci — CHANGES_REQUESTED** [review]

File: src/super_gradients/training/models/detection_models/yolo_base.py [hunk 2]
```diff
@@ -106,6 +128,31 @@ def _filter_max_predictions(self, res: List) -> List:
...
+        self,
+        conf: float = 0.001,
+        iou: float = 0.6,
+        classes: List[int] = None,
+        nms_type: NMS_Type = NMS_Type.ITERATIVE,
+        max_predictions: int = 300,
+        with_confidence: bool = True,
+        class_agnostic_nms: bool = False,
+        multi_label_per_box: bool = False,
+    ):
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/models/detection_models/yolo_base.py:140]: Fixed in 538197658b5f475582b0be058ef2e6f253c0a790

File: src/super_gradients/training/utils/detection_utils.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/detection_utils.py]: Fixed in ff370f0995797c53b43996b7570bca694a2264dd

File: src/super_gradients/training/utils/detection_utils.py [hunk 3]
```diff
@@ -284,18 +292,17 @@ def non_max_suppression(prediction, conf_thres=0.1, iou_thres=0.6, multi_label_p
 
         # Apply torch batched NMS algorithm
         boxes, scores, cls_idx = pred[:, :4], pred[:, 4], pred[:, 5]
-        idx_to_keep = torchvision.ops.boxes.batched_nms(boxes, scores, cls_idx, iou_thres)
+        if class_agnostic_nms:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/utils/detection_utils.py:295]: Well, I'm somewhat disagree.
We filter after multiplying on confidence in all cases:

```
# with_confidence = True, multi_label_per_box= False


# This evals to true and we multiply class scores by objectness
if with_confidence:
   pred[:, 5:] *= pred[:, 4:5]  # multiply objectness score with class score

```
# if multi_label_per_box evaluates to False so we go through else route:

# Detections matrix nx6 (xyxy, conf, cls)
if multi_label_per_box:  # try for all good confidence classes
    i, j = (pred[:, 5:] > conf_thres).nonzero(as_tuple=False).T
    pred = torch.cat((box[i], pred[i, j + 5, None], j[:, None].float()), 1)

else:  # best class only
    conf, j = pred[:, 5:].max(1, keepdim=True)
    pred = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres] #Here is the actual filtering
```

At the last line we take most confident class and then filter out by `(class score * objectness > conf_thres)`. So I think we indeed filter them.

**Review by shaydeci — APPROVED** [review]
LGTM
